controllers/books.py

```diff
@@ -74,25 +91,41 @@ def _route_book(is_published=True):
         response.cookies["last_course"]["path"] = "/"
 
         # Get `course info <courses table>`.
-        course = (
-            db(db.courses.id == auth.user.course_id)
-            .select(
-                db.courses.course_name,
-                db.courses.base_course,
-                db.courses.allow_pairs,
-                db.courses.downloads_enabled,
-                **cache_kwargs
-            )
-            .first()
-        )
+        course = _get_current_course(auth)
 
+        if course:
+            current_course = course.course_name
         # Ensure the base course in the URL agrees with the base course in ``course``.
         # If not, ask the user to select a course.
         if not course or course.base_course != base_course:
-            session.flash = "{} is not the course your are currently in,  switch to or add it to go there".format(
-                base_course
+            # Check to see if the user is enrolled in this base course
+            res = (
+                db(
+                    (db.user_courses.user_id == auth.user.id)
+                    & (db.user_courses.course_id == db.courses.id)
+                    & (db.courses.base_course == base_course)
+                )
+                .select()
+                .first()
             )
-            redirect(URL(c="default", f="courses"))
+            if res:
+                # change user to this course
+                db(db.auth_user.id == auth.user.id).update(course_id=res.courses.id)
+                db(db.auth_user.id == auth.user.id).update(
+                    course_name=res.courses.course_name
+                )
+                auth.user.update(course_name=res.courses.course_name)
+                auth.user.update(course_id=res.courses.id)
+                current_course = res.courses.course_name
+                course = _get_current_course(auth)
+                session.flash = (
+                    f"Automatically changing your current course to {current_course}"
```

Comment:
I'm still on Python 3.6 -- would you mind using `format` instead?

Reply:
I thought we agreed that we could move on....  I've been freely coding with f strings in the components as well.

3.9 is in beta and will be released in October.  I think should support the current release and 1 release back and be testing on the current beta so we are ready to support the current release.

Reply:
regardless, I just tried and at least in 3.6.9 f-strings seem to work fine.

Reply:
Oops, you're right. F-strings are OK (supported in Python 3.6).

How do you keep your server's Python current? I'd been using whatever ships with the LTS version of Ubuntu, and I'm hesitant to upgrade that since Python touches so many pieces of the OS.

Reply:
I use the ppa provider for python and use apt.  On my dev server I use pyenv so I can quickly change around.  No problems in any cases.

Reply:
Great, thanks! I'll try that out soon. Right now, students are taking a test on my server, so I'll have to wait for some downtime...